cmd/common/runwithspec_test.go

```diff
@@ -0,0 +1,260 @@
+		assert.Equal(r.t, r.testSpec.ExpectedTags, runtimeParams.Get(grpcruntime.ParamTags).AsStringSlice())
+	}
+	if r.testSpec.ExpectedParams != nil {
+		for k, v := range r.testSpec.ExpectedParams {
+			val, ok := paramValueMap[k]
+			require.True(r.t, ok, "expected paramValue exists for %q", k)
+			require.True(r.t, ok, "expected paramValue to be %q, got %q", v, val)
+		}
+	}
+	return nil
+}
+
+func createTempManifest(t *testing.T, manifest string) (string, error) {
+	f, err := os.CreateTemp("", "manifest")
+	if err != nil {
+		return "", fmt.Errorf("creating temp manifest file: %w", err)
+	}
+	t.Cleanup(func() {
+		os.Remove(f.Name())
+	})
+	_, err = f.WriteString(manifest)
+	if err != nil {
+		return "", fmt.Errorf("writing temp manifest: %w", err)
+	}
+	return f.Name(), nil
+}
+
+func TestRunWithSpec(t *testing.T) {
+	tests := []*testSpec{
+		{
+			Name:           "specified image on command line",
+			ExpectError:    true,
+			AdditionalArgs: []string{"trace_exec"},
+			Manifest: `
+apiVersion: 1
+kind: instance-spec
+image: trace_exec
+`,
+		},
+		{
+			Name: "invalid version",
+			Manifest: `
+apiVersion: a`,
+			ExpectError: true,
+		},
+		{
+			Name: "invalid kind",
+			Manifest: `
+apiVersion: 1
+kind: yo`,
+			ExpectError: true,
+		},
+		{
+			Name: "one spec with id, name, tags and params",
+			Manifest: `
+apiVersion: 1
+kind: instance-spec
+image: demo
+id: 00000000000000000000000000000000
+name: myinstancename
+tags:
+  - tag1
+  - tag2
+paramValues:
+  a: b
+  y: z
+  n: 1
+`,
+			ExpectedID:   "00000000000000000000000000000000",
+			ExpectedName: "myinstancename",
+			ExpectedTags: []string{"tag1", "tag2"},
+			ExpectedParams: map[string]string{
+				"a": "b",
+				"y": "z",
+				"n": "1",
+			},
+		},
+		{
+			Name: "multiple specs with detach",
+			Manifest: `
+apiVersion: 1
+kind: instance-spec
+image: demo
+---
+apiVersion: 1
+kind: instance-spec
+image: demo
+`,
+			Mode: testModeDetachOnly,
+		},
+		{
+			Name: "multiple specs without detach",
+			Manifest: `
+apiVersion: 1
+kind: instance-spec
+image: demo
+---
+apiVersion: 1
+kind: instance-spec
+image: demo
+`,
+			Mode:        testModeInteractiveOnly,
+			ExpectError: true,
+		},
+		{
+			Name: "multiple specs, one invalid",
+			Manifest: `
+apiVersion: 1
+kind: instance-spec
+image: demo
+---
+apiVersion: x
+`,
+			Mode:        testModeInteractiveOnly,
+			ExpectError: true,
+		},
+		{
+			Name: "missing image",
+			Manifest: `
+apiVersion: 1
+kind: instance-spec
+`,
+			ExpectError: true,
+		},
+		{
+			Name: "invalid id",
+			Manifest: `
+apiVersion: 1
+kind: instance-spec
+image: demo
+id: 0
+`,
+			ExpectError: true,
+		},
+		{
+			Name: "invalid name",
+			Manifest: `
+apiVersion: 1
+kind: instance-spec
+image: demo
+name: invalid-#name*
+`,
+			ExpectError: true,
+		},
+	}
+
+	environment.Environment = environment.Local
+
+	for _, tt := range tests {
+		runTest := func(detach bool) func(t *testing.T) {
+			return func(t *testing.T) {
```

Comment:
I added `t.Parallel()` and everything exploded :exploding_head:

Reply:
It seems that `setFlagsFromConfig()` has global side-effects that cause that (concurrent map read/writes). I think it's fine for the tests.